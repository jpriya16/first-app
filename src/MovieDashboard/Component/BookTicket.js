import React from "react";
import {useState} from "react";
import {updateTicket} from "./MovieService";

export default function BookTicket(movies,UpdateMovieDetails){
    const [MovieId, setMovieId] = useState('')
    const [NumberOfTickets,setNumberOfTickets] = useState(0)
    const [Book,setBook] = useState(false)
    const [message,setMessage] = useState('')


    const handleBooKTicket = () => {
        setBook(true);
        BookTicketForMovie(movies);
        setNumberOfTickets(0);
        setMovieId('');
        setBook(false)

    }

    const bookTicketWithData = async (MovieId,newMovie) => {
        return await updateTicket(MovieId,newMovie)
    }

    const BookTicketForMovie = (movies) => {
        let movie = null;
        let diff = 0;
        let total = 0;
        let booked = 0;
        for (let key in movies) {
            if(movies[key]){
                if(movies[key]._id === MovieId) {
                    console.log("movie present", movies[key])
                    movie = movies[key];
                    total = Number(movie.total_tickets);
                    booked = Number(movie.booked_tickets);
                    diff = total - (NumberOfTickets + booked)
                }
            }
        }



        if(movie == null)
            setMessage('Invalid Movie Id ' + MovieId)
        else if(diff < 0){
            setMessage('requested no of ticket is not available for Movie Id ' + MovieId)}
        else if (movie && diff >= 0) {
                let newMovie = {
                    "name": movie.name,
                    "total_tickets": total,
                    "booked_tickets": booked + NumberOfTickets
                }
                bookTicketWithData(MovieId, newMovie).then(r =>{
                    if (r){
                        UpdateMovieDetails()
                        setMessage('Tickets booked for Movie Id : '+ MovieId +'Successfully')}
                }).catch(e=>{

                        setMessage('try booking after some time')
                })
                // axios.put(`https://crudcrud.com/api/c52d5cdcc84e415e80d8f2b006cddc0b/movies/` + MovieId, newMovie).then(response => {
                //     UpdateMovieDetails()
                //     setMessage('Tickets booked for Movie Id : '+ MovieId +'Successfully')
                //     }
                // ).catch(error => {
                //     setMessage('try booking after some time')
                // })


        }
    }

    return(
        <div>
            <label>MOVIE DASHBOARD</label>
            <br/><br/>
            <label>MOVIE ID</label>
            <input data-testid ="movieId" id ="movieId" placeholder="Movie Id" value={MovieId} onChange={e=>setMovieId(e.target.value)}></input>
            <label>Number of tickets to book</label>
            <input data-testid="noOfTicket" id="noOfTicket" type="text" pattern="[0-9]*" value={NumberOfTickets} placeholder="no of tickets" onChange={e=>setNumberOfTickets(Number(e.target.value))}></input>
            <button data-testid="bookButton" id ="bookbtn" onClick={handleBooKTicket}>Book</button>
            { Book || <label>{message}</label>}
        </div>
    )
}
