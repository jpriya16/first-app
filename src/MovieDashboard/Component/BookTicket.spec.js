/* eslint-disable */

import React from "react";
import {configure, shallow} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import BookTicket from "./BookTicket";
import renderer from "react-test-renderer";
import {render, screen} from "@testing-library/react";
import userEvent from '@testing-library/user-event';

configure({ adapter: new Adapter() });

describe("Book ticket",() => {
    const mock = jest.fn();
    const movies =
         [
            {
                "_id": "123",
                "name": "Movie5",
                "total_tickets": 40,
                "booked_tickets": 10
            }
        ]
    it("should match the snapshot", ()=>{
       const component = shallow(<BookTicket movies={movies} UpdateMovieDetails={mock()} />)
        expect(component.getElements()).toMatchSnapshot();
    })

    it("should display the labels in the book ticket page", ()=>{
        const component = shallow(<BookTicket movies={movies} UpdateMovieDetails={mock()} />)
        const movieDashboardLabel = <label>MOVIE DASHBOARD</label>
        const movieIdLabel = <label>MOVIE ID</label>
        const numberOfTicketsLabel = <label>Number of tickets to book</label>
        expect(component.contains(movieDashboardLabel)).toBeTruthy();
        expect(component.contains(movieIdLabel)).toBeTruthy();
        expect(component.contains(numberOfTicketsLabel)).toBeTruthy();
    })

    it("should display button", ()=>{
        let mockComponent = renderer.create(<BookTicket movies={movies} UpdateMovieDetails={mock}/>)
        expect( mockComponent.root.findByProps({id: "bookbtn"})).toBeTruthy();
    })

    it("should dispaly book button",()=>{
        let mockComponent = renderer.create(<BookTicket movies={movies} UpdateMovieDetails={mock}/>)
        expect( mockComponent.root.findByProps({id: "bookbtn"}));
    })


    it('should change state of the input after the button is clicked', () => {
        let mock = jest.fn();
        let result = mock()

        let updateTicket = jest.fn(() => new Promise((resolve) => {
            resolve({data: true})
        }));

        render(<BookTicket movies={movies} UpdateMovieDetails={result}/>);

        const movieIdInput = screen.getByTestId("movieId");
        userEvent.type( movieIdInput, "123");
        expect(screen.getByTestId("movieId")).toHaveValue("123")

        const numberOfTicketInput = screen.getByTestId("noOfTicket");
        userEvent.type( numberOfTicketInput, "10");
        expect(screen.getByTestId("noOfTicket")).toHaveValue("10")

        const button = screen.getByTestId("bookButton")
        userEvent.click(button)

        //expect(updateTicket).toHaveBeenCalledTimes(1)            //  could not check that. throwing error.
        expect(screen.getByTestId("noOfTicket")).toHaveValue("0")
        expect(screen.getByTestId("movieId")).toHaveValue("")

    });


})
