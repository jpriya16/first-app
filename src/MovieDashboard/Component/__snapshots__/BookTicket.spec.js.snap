// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Book ticket should match the snapshot 1`] = `
Array [
  <div>
    <label>
      MOVIE DASHBOARD
    </label>
    <br />
    <br />
    <label>
      MOVIE ID
    </label>
    <input
      data-testid="movieId"
      onChange={[Function]}
      placeholder="Movie Id"
      value=""
    />
    <label>
      Number of tickets to book
    </label>
    <input
      data-testid="noOfTicket"
      onChange={[Function]}
      pattern="[0-9]*"
      placeholder="no of tickets"
      type="text"
      value={0}
    />
    <button
      data-testid="bookButton"
      id="bookbtn"
      onClick={[Function]}
    >
      Book
    </button>
  </div>,
]
`;
